subsys/emds/emds.c

```diff
@@ -253,7 +288,23 @@ int emds_prepare(void)
 	(void)emds_entries_size(&size);
 
 	rc = emds_flash_prepare(&emds_flash, size);
-	if (rc) {
+
+	/** If flash is empty after prepare, imediatley make an
+	 * extra copy of the current entries as backup before the
+	 * next EMDS event in case it should fail
+	 */
+	if (IS_ENABLED(CONFIG_EMDS_AUTO_RECOVER) && rc == 1) {
```

Comment:
If I get this correct, what we do in `emds_prepare` is:
- invalidate old entries
- if emds partition is full, then:
  - erase emds partition
  - write back to flash what we just restored
  - invalidate the written entries again

So once we reach the end of emds partition, we invalidate entries twice?

Reply:
You are right. This could have been solved in a more efficient way by extending emds_flash API to directly writing invalidated entries to the flash. I hacked it to save some bloat/time. If you think it is neccessary I can do it in the more correct way.

Reply:
So this also means that if during this procedure the power loss is triggered, we will loose the backup data as well. Considering that partition is unlikely will be so big to store all entries several times before it needs to be erased, the probability of this is quite high.

Reply:
Right, there is a chance but im not sure how likely it is. Since `emds_prepare` should be done as a intitalization step on boot, the device must be in such a state that it basically shuts down immediatley after a boot-up. The reboot must happen sometime after the flash_erase procedure begins and before the backup storing procedure ends, and in addition the emds_shutdown store procedure must fail. Do you think there is high likelyhood of this happening?

Reply:
To 100% safeguard this I guess we would need a solution with 2 partitions that can be erased seperatley (like you and Omkar mentioned for the nRF5 solution).

Reply:
> Right, there is a chance but im not sure how likely it is. Since `emds_prepare` should be done as a intitalization step on boot, the device must be in such a state that it basically shuts down immediatley after a boot-up. The reboot must happen sometime after the flash_erase procedure begins and before the backup storing procedure ends, and in addition the emds_shutdown store procedure must fail. Do you think there is high likelyhood of this happening?

Otherwise we wouldn't be asked for this feature. Yeah, it wasn't right to speculate on the probability of such things because predict such situation is almost impossible. But what we could say for sure is that the current solution is still vulnerable in some conditions and, thus, it doesn't completely solves the initial problem.

Consider also that the time to erase the whole emds partition is bigger than the time that a capacitor can give to store the data. Otherwise the capacitor size should consider erase time and double write.

We also should avoid wearing out flash with this option, because this was one of the advantages of emds.

> To 100% safeguard this I guess we would need a solution with 2 partitions that can be erased seperatley (like you and Omkar mentioned for the nRF5 solution).

I tend to think that this would be the most reliable solution. But we need to think carefully how to determine which partition to use for loading.